File: assembly/tensor.ts
let nx = 0;
let ny = 0;
let k0 = 0.0;
let dxCell = new Float64Array(0);
let dyCell = new Float64Array(0);
let dxDual = new Float64Array(0);
let dyDual = new Float64Array(0);
let epsilonX = new Float64Array(0);
let epsilonY = new Float64Array(0);
let epsilonXY = new Float64Array(0);
let epsilonXZAtNodes = new Float64Array(0);
let epsilonYZAtNodes = new Float64Array(0);
let epsilonZAtNodes = new Float64Array(0);
let epsilonXZAtEx = new Float64Array(0);
let epsilonYZAtEy = new Float64Array(0);

export const Float64Array_ID = idof<Float64Array>();

@inline
function cell(row: i32, column: i32): i32 {
  return row * nx + column;
}

@inline
function clamp(value: i32, minimum: i32, maximum: i32): i32 {
  return min(max(value, minimum), maximum);
}

export function configureTensorOperator(
  nextNx: i32,
  nextNy: i32,
  nextK0: f64,
  nextDxCell: Float64Array,
  nextDyCell: Float64Array,
  nextDxDual: Float64Array,
  nextDyDual: Float64Array,
  epsilonXXCell: Float64Array,
  epsilonYYCell: Float64Array,
  epsilonZZCell: Float64Array,
  epsilonXYCell: Float64Array,
  epsilonXZCell: Float64Array,
  epsilonYZCell: Float64Array,
): i32 {
  nx = nextNx;
  ny = nextNy;
  k0 = nextK0;
  dxCell = nextDxCell;
  dyCell = nextDyCell;
  dxDual = nextDxDual;
  dyDual = nextDyDual;
  epsilonXY = epsilonXYCell;
  epsilonX = new Float64Array((ny + 1) * nx);
  epsilonY = new Float64Array(ny * (nx + 1));
  epsilonXZAtEx = new Float64Array((ny + 1) * nx);
  epsilonYZAtEy = new Float64Array(ny * (nx + 1));
  epsilonXZAtNodes = new Float64Array((ny + 1) * (nx + 1));
  epsilonYZAtNodes = new Float64Array((ny + 1) * (nx + 1));
  epsilonZAtNodes = new Float64Array((ny + 1) * (nx + 1));

  for (let row = 0; row <= ny; row += 1) {
    const south = clamp(row - 1, 0, ny - 1);
    const north = clamp(row, 0, ny - 1);
    for (let column = 0; column < nx; column += 1) {
      const edge = row * nx + column;
      epsilonX[edge] = 0.5 * (epsilonXXCell[cell(south, column)] + epsilonXXCell[cell(north, column)]);
      epsilonXZAtEx[edge] = 0.5 * (epsilonXZCell[cell(south, column)] + epsilonXZCell[cell(north, column)]);
    }
  }
  for (let row = 0; row < ny; row += 1) {
    for (let column = 0; column <= nx; column += 1) {
      const west = clamp(column - 1, 0, nx - 1);
      const east = clamp(column, 0, nx - 1);
      const edge = row * (nx + 1) + column;
      epsilonY[edge] = 0.5 * (epsilonYYCell[cell(row, west)] + epsilonYYCell[cell(row, east)]);
      epsilonYZAtEy[edge] = 0.5 * (epsilonYZCell[cell(row, west)] + epsilonYZCell[cell(row, east)]);
    }
  }
  for (let row = 0; row <= ny; row += 1) {
    const south = clamp(row - 1, 0, ny - 1);
    const north = clamp(row, 0, ny - 1);
    for (let column = 0; column <= nx; column += 1) {
      const west = clamp(column - 1, 0, nx - 1);
      const east = clamp(column, 0, nx - 1);
      const node = row * (nx + 1) + column;
      const southwest = cell(south, west);
      const southeast = cell(south, east);
      const northwest = cell(north, west);
      const northeast = cell(north, east);
      epsilonZAtNodes[node] = 0.25 * (epsilonZZCell[southwest] + epsilonZZCell[southeast] + epsilonZZCell[northwest] + epsilonZZCell[northeast]);
      epsilonXZAtNodes[node] = 0.25 * (epsilonXZCell[southwest] + epsilonXZCell[southeast] + epsilonXZCell[northwest] + epsilonXZCell[northeast]);
      epsilonYZAtNodes[node] = 0.25 * (epsilonYZCell[southwest] + epsilonYZCell[southeast] + epsilonYZCell[northwest] + epsilonYZCell[northeast]);
    }
  }
  return 2 * (ny * (nx + 1) + (ny + 1) * nx);
}

export function applyTensorOperator(input: Float64Array): Float64Array {
  const hxSize = ny * (nx + 1);
  const hySize = (ny + 1) * nx;
  const physicalSize = 2 * (hxSize + hySize);
  assert(input.length == physicalSize);
  const output = new Float64Array(physicalSize);
  const exOffset = 0;
  const eyOffset = hySize;
  const hxOffset = hySize + hxSize;
  const hyOffset = hySize + 2 * hxSize;
  const outputExOffset = 0;
  const outputEyOffset = hySize;
  const outputHxOffset = hySize + hxSize;
  const outputHyOffset = hySize + 2 * hxSize;
  const exCell = new Float64Array(nx * ny);
  const eyCell = new Float64Array(nx * ny);
  const exNode = new Float64Array((nx + 1) * (ny + 1));
  const eyNode = new Float64Array((nx + 1) * (ny + 1));
  const ezNode = new Float64Array((nx + 1) * (ny + 1));
  const longitudinalE = new Float64Array(nx * ny);

  for (let row = 0; row < ny; row += 1) {
    for (let column = 0; column < nx; column += 1) {
      const index = cell(row, column);
      exCell[index] = 0.5 * (input[exOffset + row * nx + column] + input[exOffset + (row + 1) * nx + column]);
      eyCell[index] = 0.5 * (input[eyOffset + row * (nx + 1) + column] + input[eyOffset + row * (nx + 1) + column + 1]);
      longitudinalE[index] = (
        input[exOffset + (row + 1) * nx + column] - input[exOffset + row * nx + column]
      ) / dyCell[row] - (
        input[eyOffset + row * (nx + 1) + column + 1] - input[eyOffset + row * (nx + 1) + column]
      ) / dxCell[column];
    }
  }

  for (let row = 0; row <= ny; row += 1) {
    for (let column = 0; column <= nx; column += 1) {
      const node = row * (nx + 1) + column;
      const westEx = input[exOffset + row * nx + clamp(column - 1, 0, nx - 1)];
      const eastEx = input[exOffset + row * nx + clamp(column, 0, nx - 1)];
      const southEy = input[eyOffset + clamp(row - 1, 0, ny - 1) * (nx + 1) + column];
      const northEy = input[eyOffset + clamp(row, 0, ny - 1) * (nx + 1) + column];
      exNode[node] = 0.5 * (westEx + eastEx);
      eyNode[node] = 0.5 * (southEy + northEy);
      const southHx = row > 0 ? input[hxOffset + (row - 1) * (nx + 1) + column] : 0.0;
      const northHx = row < ny ? input[hxOffset + row * (nx + 1) + column] : 0.0;
      const westHy = column > 0 ? input[hyOffset + row * nx + column - 1] : 0.0;
      const eastHy = column < nx ? input[hyOffset + row * nx + column] : 0.0;
      const displacementZ = (northHx - southHx) / dyDual[row] - (eastHy - westHy) / dxDual[column];
      ezNode[node] = (displacementZ - epsilonXZAtNodes[node] * exNode[node] - epsilonYZAtNodes[node] * eyNode[node]) / epsilonZAtNodes[node];
    }
  }

  for (let row = 0; row <= ny; row += 1) {
    for (let column = 0; column < nx; column += 1) {
      const edge = row * nx + column;
      output[outputExOffset + edge] = (ezNode[row * (nx + 1) + column + 1] - ezNode[row * (nx + 1) + column]) / (k0 * dxCell[column]) - k0 * input[hyOffset + edge];
      const south = cell(clamp(row - 1, 0, ny - 1), column);
      const north = cell(clamp(row, 0, ny - 1), column);
      const crossY = 0.5 * (epsilonXY[south] * eyCell[south] + epsilonXY[north] * eyCell[north]);
      const ezAtEx = 0.5 * (ezNode[row * (nx + 1) + column] + ezNode[row * (nx + 1) + column + 1]);
      const displacementX = epsilonX[edge] * input[exOffset + edge] + crossY + epsilonXZAtEx[edge] * ezAtEx;
      const southLongitudinal = row > 0 ? longitudinalE[cell(row - 1, column)] : 0.0;
      const northLongitudinal = row < ny ? longitudinalE[cell(row, column)] : 0.0;
      output[outputHyOffset + edge] = -(northLongitudinal - southLongitudinal) / (k0 * dyDual[row]) - k0 * displacementX;
    }
  }

  for (let row = 0; row < ny; row += 1) {
    for (let column = 0; column <= nx; column += 1) {
      const edge = row * (nx + 1) + column;
      output[outputEyOffset + edge] = (ezNode[(row + 1) * (nx + 1) + column] - ezNode[row * (nx + 1) + column]) / (k0 * dyCell[row]) + k0 * input[hxOffset + edge];
      const west = cell(row, clamp(column - 1, 0, nx - 1));
      const east = cell(row, clamp(column, 0, nx - 1));
      const crossX = 0.5 * (epsilonXY[west] * exCell[west] + epsilonXY[east] * exCell[east]);
      const ezAtEy = 0.5 * (ezNode[row * (nx + 1) + column] + ezNode[(row + 1) * (nx + 1) + column]);
      const displacementY = epsilonY[edge] * input[eyOffset + edge] + crossX + epsilonYZAtEy[edge] * ezAtEy;
      const westLongitudinal = column > 0 ? longitudinalE[cell(row, column - 1)] : 0.0;
      const eastLongitudinal = column < nx ? longitudinalE[cell(row, column)] : 0.0;
      output[outputHxOffset + edge] = -(eastLongitudinal - westLongitudinal) / (k0 * dxDual[column]) + k0 * displacementY;
    }
  }
  return output;
}

@inline
function dot(first: Float64Array, second: Float64Array): f64 {
  let sum = 0.0;
  for (let index = 0; index < first.length; index += 1) sum += first[index] * second[index];
  return sum;
}

@inline
function norm(values: Float64Array): f64 {
  return Math.sqrt(dot(values, values));
}

@inline
function addScaled(target: Float64Array, source: Float64Array, factor: f64): void {
  for (let index = 0; index < target.length; index += 1) target[index] += factor * source[index];
}

function applyShifted(vector: Float64Array, shift: f64): Float64Array {
  const output = applyTensorOperator(vector);
  addScaled(output, vector, -shift);
  return output;
}

export function solveShiftedTensorSystem(
  rightHandSide: Float64Array,
  shift: f64,
  maximumIterations: i32 = 180,
  relativeTolerance: f64 = 1e-5,
): Float64Array {
  const size = rightHandSide.length;
  const solution = new Float64Array(size);
  let residual = rightHandSide.slice();
  const shadow = residual.slice();
  const direction = new Float64Array(size);
  let operatorDirection = new Float64Array(size);
  let rhoPrevious = 1.0;
  let alpha = 1.0;
  let omega = 1.0;
  const tolerance = relativeTolerance * max(norm(rightHandSide), 1.0);

  for (let iteration = 0; iteration < maximumIterations; iteration += 1) {
    const rho = dot(shadow, residual);
    if (Math.abs(rho) < 1e-30) break;
    const beta = (rho / rhoPrevious) * (alpha / omega);
    for (let index = 0; index < size; index += 1) {
      direction[index] = residual[index] + beta * (direction[index] - omega * operatorDirection[index]);
    }
    operatorDirection = applyShifted(direction, shift);
    const denominator = dot(shadow, operatorDirection);
    if (Math.abs(denominator) < 1e-30) break;
    alpha = rho / denominator;
    const intermediate = residual.slice();
    addScaled(intermediate, operatorDirection, -alpha);
    if (norm(intermediate) <= tolerance) {
      addScaled(solution, direction, alpha);
      return solution;
    }
    const operatorIntermediate = applyShifted(intermediate, shift);
    const omegaDenominator = dot(operatorIntermediate, operatorIntermediate);
    if (omegaDenominator < 1e-30) break;
    omega = dot(operatorIntermediate, intermediate) / omegaDenominator;
    addScaled(solution, direction, alpha);
    addScaled(solution, intermediate, omega);
    residual = intermediate;
    addScaled(residual, operatorIntermediate, -omega);
    if (norm(residual) <= tolerance) return solution;
    if (Math.abs(omega) < 1e-30) break;
    rhoPrevious = rho;
  }
  return solution;
}

File: rust/bend-solver/src/tensor.rs
#[derive(Debug)]
pub struct TensorOperator {
    nx: usize,
    ny: usize,
    k0: f64,
    dx_cell: Vec<f64>,
    dy_cell: Vec<f64>,
    dx_dual: Vec<f64>,
    dy_dual: Vec<f64>,
    epsilon_x: Vec<f64>,
    epsilon_y: Vec<f64>,
    epsilon_xy: Vec<f64>,
    epsilon_xz_nodes: Vec<f64>,
    epsilon_yz_nodes: Vec<f64>,
    epsilon_z_nodes: Vec<f64>,
    epsilon_xz_ex: Vec<f64>,
    epsilon_yz_ey: Vec<f64>,
}

impl TensorOperator {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        nx: usize,
        ny: usize,
        k0: f64,
        dx_cell: Vec<f64>,
        dy_cell: Vec<f64>,
        dx_dual: Vec<f64>,
        dy_dual: Vec<f64>,
        epsilon_xx_cell: Vec<f64>,
        epsilon_yy_cell: Vec<f64>,
        epsilon_zz_cell: Vec<f64>,
        epsilon_xy: Vec<f64>,
        epsilon_xz_cell: Vec<f64>,
        epsilon_yz_cell: Vec<f64>,
    ) -> Result<Self, u32> {
        let cells = nx * ny;
        if nx == 0
            || ny == 0
            || !k0.is_finite()
            || k0 <= 0.0
            || dx_cell.len() != nx
            || dy_cell.len() != ny
            || dx_dual.len() != nx + 1
            || dy_dual.len() != ny + 1
            || [epsilon_xx_cell.len(), epsilon_yy_cell.len(), epsilon_zz_cell.len(), epsilon_xy.len(), epsilon_xz_cell.len(), epsilon_yz_cell.len()]
                .iter()
                .any(|length| *length != cells)
            || dx_cell.iter().chain(&dy_cell).chain(&dx_dual).chain(&dy_dual).any(|value| !value.is_finite() || *value <= 0.0)
        {
            return Err(20);
        }
        let mut epsilon_x = vec![0.0; (ny + 1) * nx];
        let mut epsilon_y = vec![0.0; ny * (nx + 1)];
        let mut epsilon_xz_ex = vec![0.0; (ny + 1) * nx];
        let mut epsilon_yz_ey = vec![0.0; ny * (nx + 1)];
        let mut epsilon_xz_nodes = vec![0.0; (ny + 1) * (nx + 1)];
        let mut epsilon_yz_nodes = vec![0.0; (ny + 1) * (nx + 1)];
        let mut epsilon_z_nodes = vec![0.0; (ny + 1) * (nx + 1)];

        for row in 0..=ny {
            let south = row.saturating_sub(1).min(ny - 1);
            let north = row.min(ny - 1);
            for column in 0..nx {
                let edge = row * nx + column;
                epsilon_x[edge] = 0.5 * (epsilon_xx_cell[south * nx + column] + epsilon_xx_cell[north * nx + column]);
                epsilon_xz_ex[edge] = 0.5 * (epsilon_xz_cell[south * nx + column] + epsilon_xz_cell[north * nx + column]);
            }
        }
        for row in 0..ny {
            for column in 0..=nx {
                let west = column.saturating_sub(1).min(nx - 1);
                let east = column.min(nx - 1);
                let edge = row * (nx + 1) + column;
                epsilon_y[edge] = 0.5 * (epsilon_yy_cell[row * nx + west] + epsilon_yy_cell[row * nx + east]);
                epsilon_yz_ey[edge] = 0.5 * (epsilon_yz_cell[row * nx + west] + epsilon_yz_cell[row * nx + east]);
            }
        }
        for row in 0..=ny {
            let south = row.saturating_sub(1).min(ny - 1);
            let north = row.min(ny - 1);
            for column in 0..=nx {
                let west = column.saturating_sub(1).min(nx - 1);
                let east = column.min(nx - 1);
                let node = row * (nx + 1) + column;
                let indices = [south * nx + west, south * nx + east, north * nx + west, north * nx + east];
                epsilon_z_nodes[node] = 0.25 * indices.iter().map(|index| epsilon_zz_cell[*index]).sum::<f64>();
                epsilon_xz_nodes[node] = 0.25 * indices.iter().map(|index| epsilon_xz_cell[*index]).sum::<f64>();
                epsilon_yz_nodes[node] = 0.25 * indices.iter().map(|index| epsilon_yz_cell[*index]).sum::<f64>();
            }
        }
        Ok(Self {
            nx,
            ny,
            k0,
            dx_cell,
            dy_cell,
            dx_dual,
            dy_dual,
            epsilon_x,
            epsilon_y,
            epsilon_xy,
            epsilon_xz_nodes,
            epsilon_yz_nodes,
            epsilon_z_nodes,
            epsilon_xz_ex,
            epsilon_yz_ey,
        })
    }

    pub fn physical_size(&self) -> usize {
        2 * (self.ny * (self.nx + 1) + (self.ny + 1) * self.nx)
    }

    pub fn apply(&self, input: &[f64]) -> Result<Vec<f64>, u32> {
        let nx = self.nx;
        let ny = self.ny;
        let hx_size = ny * (nx + 1);
        let hy_size = (ny + 1) * nx;
        let physical_size = 2 * (hx_size + hy_size);
        if input.len() != physical_size {
            return Err(21);
        }
        let ex_offset = 0;
        let ey_offset = hy_size;
        let hx_offset = hy_size + hx_size;
        let hy_offset = hy_size + 2 * hx_size;
        let mut output = vec![0.0; physical_size];
        let mut ex_cell = vec![0.0; nx * ny];
        let mut ey_cell = vec![0.0; nx * ny];
        let mut ex_node = vec![0.0; (nx + 1) * (ny + 1)];
        let mut ey_node = vec![0.0; (nx + 1) * (ny + 1)];
        let mut ez_node = vec![0.0; (nx + 1) * (ny + 1)];
        let mut longitudinal_e = vec![0.0; nx * ny];

        for row in 0..ny {
            for column in 0..nx {
                let cell = row * nx + column;
                ex_cell[cell] = 0.5 * (input[ex_offset + row * nx + column] + input[ex_offset + (row + 1) * nx + column]);
                ey_cell[cell] = 0.5 * (input[ey_offset + row * (nx + 1) + column] + input[ey_offset + row * (nx + 1) + column + 1]);
                longitudinal_e[cell] = (input[ex_offset + (row + 1) * nx + column] - input[ex_offset + row * nx + column]) / self.dy_cell[row]
                    - (input[ey_offset + row * (nx + 1) + column + 1] - input[ey_offset + row * (nx + 1) + column]) / self.dx_cell[column];
            }
        }
        for row in 0..=ny {
            for column in 0..=nx {
                let node = row * (nx + 1) + column;
                let west = column.saturating_sub(1).min(nx - 1);
                let east = column.min(nx - 1);
                let south = row.saturating_sub(1).min(ny - 1);
                let north = row.min(ny - 1);
                ex_node[node] = 0.5 * (input[ex_offset + row * nx + west] + input[ex_offset + row * nx + east]);
                ey_node[node] = 0.5 * (input[ey_offset + south * (nx + 1) + column] + input[ey_offset + north * (nx + 1) + column]);
                let south_hx = if row > 0 { input[hx_offset + (row - 1) * (nx + 1) + column] } else { 0.0 };
                let north_hx = if row < ny { input[hx_offset + row * (nx + 1) + column] } else { 0.0 };
                let west_hy = if column > 0 { input[hy_offset + row * nx + column - 1] } else { 0.0 };
                let east_hy = if column < nx { input[hy_offset + row * nx + column] } else { 0.0 };
                let displacement_z = (north_hx - south_hx) / self.dy_dual[row] - (east_hy - west_hy) / self.dx_dual[column];
                ez_node[node] = (displacement_z - self.epsilon_xz_nodes[node] * ex_node[node] - self.epsilon_yz_nodes[node] * ey_node[node])
                    / self.epsilon_z_nodes[node];
            }
        }
        for row in 0..=ny {
            for column in 0..nx {
                let edge = row * nx + column;
                output[ex_offset + edge] = (ez_node[row * (nx + 1) + column + 1] - ez_node[row * (nx + 1) + column]) / (self.k0 * self.dx_cell[column])
                    - self.k0 * input[hy_offset + edge];
                let south = row.saturating_sub(1).min(ny - 1) * nx + column;
                let north = row.min(ny - 1) * nx + column;
                let cross_y = 0.5 * (self.epsilon_xy[south] * ey_cell[south] + self.epsilon_xy[north] * ey_cell[north]);
                let ez_at_ex = 0.5 * (ez_node[row * (nx + 1) + column] + ez_node[row * (nx + 1) + column + 1]);
                let displacement_x = self.epsilon_x[edge] * input[ex_offset + edge] + cross_y + self.epsilon_xz_ex[edge] * ez_at_ex;
                let south_longitudinal = if row > 0 { longitudinal_e[(row - 1) * nx + column] } else { 0.0 };
                let north_longitudinal = if row < ny { longitudinal_e[row * nx + column] } else { 0.0 };
                output[hy_offset + edge] = -(north_longitudinal - south_longitudinal) / (self.k0 * self.dy_dual[row]) - self.k0 * displacement_x;
            }
        }
        for row in 0..ny {
            for column in 0..=nx {
                let edge = row * (nx + 1) + column;
                output[ey_offset + edge] = (ez_node[(row + 1) * (nx + 1) + column] - ez_node[row * (nx + 1) + column]) / (self.k0 * self.dy_cell[row])
                    + self.k0 * input[hx_offset + edge];
                let west = row * nx + column.saturating_sub(1).min(nx - 1);
                let east = row * nx + column.min(nx - 1);
                let cross_x = 0.5 * (self.epsilon_xy[west] * ex_cell[west] + self.epsilon_xy[east] * ex_cell[east]);
                let ez_at_ey = 0.5 * (ez_node[row * (nx + 1) + column] + ez_node[(row + 1) * (nx + 1) + column]);
                let displacement_y = self.epsilon_y[edge] * input[ey_offset + edge] + cross_x + self.epsilon_yz_ey[edge] * ez_at_ey;
                let west_longitudinal = if column > 0 { longitudinal_e[row * nx + column - 1] } else { 0.0 };
                let east_longitudinal = if column < nx { longitudinal_e[row * nx + column] } else { 0.0 };
                output[hx_offset + edge] = -(east_longitudinal - west_longitudinal) / (self.k0 * self.dx_dual[column]) + self.k0 * displacement_y;
            }
        }
        Ok(output)
    }
}
